src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -45,5 +32,227 @@ private static IEnumerable<TSource> PrependIterator<TSource>(IEnumerable<TSource
+                _sources = sources;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new ConcatNIterator<TSource>(_sources);
+            }
+
+            protected override IEnumerable<TSource> GetEnumerable(int index)
+            {
+                return index < _sources.Length ? _sources[index] : null;
+            }
+        }
+
+        private abstract class ConcatIterator<TSource> : Iterator<TSource>, IIListProvider<TSource>
+        {
+            private IEnumerator<TSource> _enumerator;
+
+            public override void Dispose()
+            {
+                if (_enumerator != null)
+                {
+                    _enumerator.Dispose();
+                    _enumerator = null;
+                }
+                base.Dispose();
+            }
+
+            protected abstract IEnumerable<TSource> GetEnumerable(int index);
+
+            public override bool MoveNext()
+            {
+                if (state == 1)
+                {
+                    _enumerator = GetEnumerable(0).GetEnumerator();
+                    state = 2;
+                }
+
+                if (state > 1)
+                {
+                    while (true)
+                    {
+                        if (_enumerator.MoveNext())
+                        {
+                            current = _enumerator.Current;
+                            return true;
+                        }
+
+                        Debug.Assert(state < int.MaxValue);
+                        IEnumerable<TSource> next = GetEnumerable(state++ - 1);
+                        if (next != null)
+                        {
+                            _enumerator.Dispose();
+                            _enumerator = next.GetEnumerator();
+                            continue;
+                        }
+
+                        Dispose();
+                        break;
+                    }
+                }
+
+                return false;
+            }
+
+            public TSource[] ToArray()
+            {
+                return EnumerableHelpers.ToArray(this);
+            }
+
+            public List<TSource> ToList()
+            {
+                var list = new List<TSource>();
+                for (int i = 0; ; i++)
+                {
+                    IEnumerable<TSource> source = GetEnumerable(i);
+                    if (source == null) break;
+
+                    list.AddRange(source);
+                }
+                return list;
+            }
+
+            public int GetCount(bool onlyIfCheap)
+            {
+                int count = 0;
+                for (int i = 0; ; i++)
+                {
+                    IEnumerable<TSource> source = GetEnumerable(i);
+                    if (source == null) break;
+
+                    ICollection<TSource> c = source as ICollection<TSource>;
```

Comment:
I didn't consider this to be a strong case for `IIListProvider` at all, for similar reasons (but did if the sources are all lists, which was my main thing about my stab at it). Of course since my plan was to have separate implementations when all inputs are lists, I'd already set things so to make cheap counts less likely, but having some of the most common cases where they are handled elsewhere.